File: big-o/2-3.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func main() {
	//someSlice := []int{5, 4, 0, 2, 1}
	someRand := rand.New(rand.NewSource(time.Now().Unix()))
	someSlice := someRand.Perm(100000)
	minN2(someSlice)
	minN(someSlice)
}

func minN2(arr []int) int {
	defer timeTrack(time.Now(), "N^2")
	min := arr[0]
	smallest := true
	for _, e1 := range arr {
		smallest = true
		for _, e2 := range arr {
			if e1 > e2 {
				smallest = false
			}
		}
		if smallest {
			min = e1
		}
	}

	return min
}

func minN(arr []int) int {
	defer timeTrack(time.Now(), "N")
	min := arr[0]
	for _, e1 := range arr {
		if e1 < min {
			min = e1
		}
	}

	return min
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	fmt.Println("%s took %s", name, elapsed)
}

File: big-o/2-3.py
import time
from random import randrange


def find_min_n2(arr):
    min = arr[0]
    for i in arr:
        smallest = True
        for j in arr:
            if (i > j):
                smallest = False

        if smallest:
            min = i
    return min


def find_min_n(arr):
    min = arr[0]
    for i in arr:
        if (i < min):
            min = i
    return min


for listSize in range(1000, 10001, 1000):
    alist = [randrange(100000) for x in range(listSize)]
    start = time.time()
    print find_min_n2(alist)
    end = time.time()
    print "Size: %d time: %f" % (listSize, end - start)

for listSize in range(1000, 10001, 1000):
    alist = [randrange(100000) for x in range(listSize)]
    start = time.time()
    print find_min_n(alist)
    end = time.time()
    print "Size: %d time: %f" % (listSize, end - start)
